Added new rule to find as type assertions and replace with satisfies

jpolavar opened the pull request:
Closes #116

carlansley commented on src/no-type-assertion-as.ts:
```diff
@@ -0,0 +1,48 @@
+            const sourceCode = context.sourceCode;
+            const typeAnnotation = sourceCode.getText(node.typeAnnotation);
+            const expression = sourceCode.getText(node.expression);
+            return fixer.replaceText(node, `${expression} satisfies ${typeAnnotation}`);
```
I have no idea if it's possible (maybe @le-cong knows) but ideally we'd only do this substitution if using `satisfies` is ok with the compiler.

carlansley replied:
also one other situation I can think of - `x as unknown as Y` should be an error but `x satisfies unknown satisfies Y` isn't going to work as a fix.

le-cong replied:
i haven't get a chance to look into this, but i feel that this rule is more complex than what #116 describes. it might be necessary to use typescript's api to compare the source/target typing in details in order to tell which the replacement is legit. we don't necessarily want to replace all `as` with `satisfies` i am afraid ...

carlansley replied:
yeah i don't think the rule should introduce compiler errors.  if we can't use the TS API to work it out, might be best not to auto-fix at all.

jpolavar replied:
Added to compare source/target typing to ensure type compatibility, verifying whether the originalType satisfies the constraints of the targetType and removed auto fix

carlansley commented on src/no-type-assertion-as.ts:
```diff
@@ -0,0 +1,60 @@
+            node.typeAnnotation.type === AST_NODE_TYPES.TSIntersectionType) &&
+          node.typeAnnotation.types.some((type) => type.type === AST_NODE_TYPES.TSUnknownKeyword);
+
+        if (
```
this is a complex bit of logic I don't quite understand, if there was just one `context.report()` it would help.  also maybe some more `isSomethingSomething` booleans to help readability.

carlansley reviewed (changes requested):
branch needs to be updated from main

carlansley reviewed (changes requested):
To @le-cong's point, I think we shouldn't do this rule at all if we can't generally understand when `as` is replaceable, and fix at least a majority of the remaining issues automatically.

jpolavar commented:
> To @le-cong's point, I think we shouldn't do this rule at all if we can't generally understand when `as` is replaceable, and fix at least a majority of the remaining issues automatically. 

Removed the auto-fix and report the error and will add parserServices and checker are used to access TypeScript's type-checking capabilities for source and target and determine whether to report or not

jpolavar commented:
> > To @le-cong's point, I think we shouldn't do this rule at all if we can't generally understand when `as` is replaceable, and fix at least a majority of the remaining issues automatically.
> 
> @jpolavar what is the end result of this on some of our bigger services? (maybe Slack me the results)

I updated it to detect as type assertions and ensure the original type and target type are assignable using TypeScript's type checker. Only then report a violation if the types are assignable, as is unnecessary in such case. Since we have added type checker, i will enhance to fix as well instead of just reporting since it would be lot to manually fix.
